### PR TITLE
Thirdperson Camera Improvements + Free movement while taunting.

### PR DESCRIPTION
**Pull request by PYROTEKNIK**
Player can now move around while playing certain taunts. Switches to server's third person camera instead of the built in taunt camera when taunting. Third person camera now toggles smoothly.

Third person camera moves to player shoulder while carrying aimable weapon.
Player can freely rotate the camera around their character with middle click if no existing bind is using it or a bind to +swamp_freecam

https://user-images.githubusercontent.com/3793593/126064023-28bab993-266f-4691-ab20-5e9f30fde9f4.mp4

### CODE REVIEW

**Comment by swampservers** [comment]
Why would you want free movement while taunting... wouldn't the feet slide along the ground like you see on fatkid when people use act commands?

**Comment by PYROTEKNIK** [comment]
Some do that, but some blend with the walk animation.  I kinda feel it's just better to not use it at all since it's a whole separate third person camera for one thing, and it's literally just a client side hook stopping you from moving

File: lua/autorun/client/thirdperson.lua [hunk 1]
```diff
@@ -21,28 +21,157 @@ hook.Add("Think", "ThirdPersonToggler", function()
...
+        Thirdperson_Lerp = 0
+        Thirdperson_LerpSide = 0
+
+        return false
+    end
 
     return THIRDPERSON or wc == "weapon_fists" or LocalPlayer():HasWeapon("weapon_goohulk") or (wc == "weapon_garfield" and IsValid(LocalPlayer()) and not LocalPlayer():InVehicle())
 end
 
+local cwep = {}
```
Comment (swampservers) [inline lua/autorun/client/thirdperson.lua]:
```
local cwep = {
weapon_pistol = true,
weapon_357 = true,
etc
}
```

File: gamemodes/cinema/gamemode/player_class/player_lobby.lua [hunk 2]
```diff
@@ -0,0 +1,97 @@
+AddCSLuaFile()
+DEFINE_BASECLASS( "player_default" )
```
Comment (swampservers) [inline gamemodes/cinema/gamemode/player_class/player_lobby.lua]: isn't this file from the other repo

**Comment by swampservers** [comment]
What currently happens when you taunt? it's still thirdperson right? 

It should still lock you in place if the legs would slide on the ground otherwise, but it can allow movement if the legs still move.

File: gamemodes/cinema/gamemode/player_class/player_lobby.lua (hunk 2 above)
Comment (PYROTEKNIK) [reply gamemodes/cinema/gamemode/player_class/player_lobby.lua]: It's a modified version of the stock cinema one

**Comment by PYROTEKNIK** [comment]
The third person camera the server uses is enabled when taunting in the same manner as it is when fists are equipped. I can add a check for specific sequence activity, for movement, I just figured it wouldn't matter much

File: gamemodes/cinema/gamemode/player_class/player_lobby.lua [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+--
+-- Allow player class to create move
+--
+
+
+function PLAYER:CreateMove( cmd )
+
+	--if ( self.TauntCam:CreateMove( cmd, self.Player, self.Player:IsPlayingTaunt() ) ) then return true end
+	local ply = self.Player 
+	if(IsLockedByTaunting(ply))then
```
Comment (swampservers) [inline gamemodes/cinema/gamemode/player_class/player_lobby.lua]: Stop making if statements like this please

File: lua/autorun/client/thirdperson.lua [hunk 4]
```diff
@@ -21,28 +21,194 @@ hook.Add("Think", "ThirdPersonToggler", function()
...
+	ACT_GMOD_TAUNT_DANCE = true,
+	ACT_GMOD_TAUNT_ROBOT = true,
+	ACT_GMOD_TAUNT_PERSISTENCE = true,
+	ACT_GMOD_TAUNT_LAUGH = true,
+	ACT_GMOD_TAUNT_CHEER = true
+}
+
+
+function IsLockedByTaunting(ply)
+    if ply:IsPlayingTaunt() then
```
Comment (swampservers) [inline lua/autorun/client/thirdperson.lua]: `return ply:IsPlayingTaunt() and LockTaunts[ply:GetSequenceActivityName(ply:GetLayerSequence(5))]`

File: gamemodes/cinema/gamemode/player_class/player_lobby.lua (hunk 2 above)
Comment (swampservers) [reply gamemodes/cinema/gamemode/player_class/player_lobby.lua]: I see, I already have one of these files in my private repo, so, please just remove it from this PR and i'll incorporate the changes from here manually

File: gamemodes/cinema/gamemode/player_class/player_lobby.lua (hunk 3 above)
Comment (PYROTEKNIK) [reply gamemodes/cinema/gamemode/player_class/player_lobby.lua]: sorry about that i'm trying to kick the habit

**Comment by swamproyal** [comment]
This needs to be closed now

**Pull request closed without merging**